pkg/model/config/config.go

```diff
@@ -113,3 +120,87 @@ func (r GVK) isEqualTo(other GVK) bool {
 		r.Version == other.Version &&
 		r.Kind == other.Kind
 }
+
+func (c Config) Marshal() ([]byte, error) {
+	// Ignore extra fields at first.
+	cfg := c
+	cfg.ExtraFields = nil
+	content, err := yaml.Marshal(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling project configuration: %v", err)
+	}
+	// Empty config strings are "{}" due to the map field.
+	if strings.TrimSpace(string(content)) == "{}" {
+		content = []byte{}
+	}
+	// Append extra fields to put them at the config's bottom, unless the
+	// project is v1 which does not support extra fields.
+	if !cfg.IsV1() && len(c.ExtraFields) != 0 {
+		extraFieldBytes, err := yaml.Marshal(Config{ExtraFields: c.ExtraFields})
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling project configuration extra fields: %v", err)
+		}
+		content = append(content, extraFieldBytes...)
+	}
+	return content, nil
+}
+
+func Unmarshal(in []byte, out *Config) error {
+	if err := yaml.UnmarshalStrict(in, out); err != nil {
+		return fmt.Errorf("error unmarshalling project configuration: %v", err)
+	}
+	// v1 projects do not support extra fields.
+	if out.IsV1() {
+		out.ExtraFields = nil
+	}
+	return nil
+}
+
+// EncodeExtraFields encodes extraFieldsObj in c. This method is intended to
+// be used for custom configuration objects.
+func (c *Config) EncodeExtraFields(key string, extraFieldsObj interface{}) error {
+	// Short-circuit v1
+	if c.IsV1() {
+		return fmt.Errorf("v1 project configs do not have extra fields")
+	}
+	// Ignore zero'd objects so we don't add empty map keys.
+	if reflect.ValueOf(extraFieldsObj).IsZero() {
+		return nil
+	}
```

Comment:
Do we really need this check? If the field was added to the ExtraFields map, I think we should encode it, the default doesn't need to be the zero value (for example versioning defaulting to 0 doesn't make much sense in case some version wants to be stored).

It has the added benefit of removing `reflect` which I think is nice.